src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,165 @@
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.value = value;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        Node<T> newNode;
+        if (size == 0) {
+            newNode = new Node<>(null, value, null);
+            head = newNode;
+            tail = newNode;
+            size++;
+            return true;
+        }
+        newNode = new Node<>(tail, value, null);
+        tail.next = newNode;
+        tail = newNode;
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        Node<T> newNode;
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+            size++;
+            return;
+        }
+        Node<T> currentNode = searchByIndex(index);
+        newNode = new Node<T>(currentNode.prev, value, currentNode);
+        currentNode.prev.next = newNode;
+        currentNode.prev = newNode;
+        size++;
+        return;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return searchByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> currentNode = searchByIndex(index);
+        T output = currentNode.value;
+        currentNode.value = value;
+        return output;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> output;
+        if (size == 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == 0) {
+            output = head;
+            if (head.next != null) {
+                head.next.prev = null;
+            }
+            head = head.next;
+            output.next = null;
+            size--;
+            return output.value;
+        }
+        if (index == size - 1) {
+            output = tail;
+            if (tail.prev != null) {
+                tail.prev.next = null;
+            }
+            tail = tail.prev;
+            output.prev = null;
+            size--;
+            return output.value;
+        }
+        Node<T> currentNode = searchByIndex(index);
+        output = currentNode;
+        currentNode.prev.next = currentNode.next;
+        currentNode.next.prev = currentNode.prev;
+        size--;
+        return output.value;
```

Comment:
let's all logic after line:  `Node<T> currentNode = searchByIndex(index);` to the separate method delete()/unlink() and use this method in the methods remove by index and remove by value

```suggestion
        Node<T> currentNode = searchByIndex(index);
        T element = currentNode.value;
        Node<T> prev = currentNode.prev;
        Node<T> next = currentNode.next;
        if (prev == null) {
            head = next;
        } else {
            prev.next = next;
        }
        if (next == null) {
            tail = prev;
        } else {
            next.prev = prev;
        }
        size--;
        return element;
```

Reply:
done